### PR TITLE
chore: moved store validation to language

### PR DESCRIPTION
**Pull request by d-jeffery**
## References

- closes https://github.com/openfga/language/issues/278 

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/js/tests/validate-rules.test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import { describe, expect, it } from "@jest/globals";
+import { Validate } from "../validator/validate-rules";
+
+describe("validation rules", () => {
+  describe("Rule 'Object'", () => {
+    it("should pass '<object>:<id>'", () => {
+      expect(Validate.object("document:1")).toBeTruthy();
+    });
+
+    it("should fail if no ':' delimiter", () => {
```
Comment (rhamzeh) [inline pkg/js/tests/validate-rules.test.ts:10]: Can you add more test cases?

These should fail
* document::1
* :document:1
* document:1:
* document#1
* document:1#relation

File: pkg/js/tests/validate-rules.test.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  describe("Rule 'User'", () => {
+    it("shoud pass for 'UserSet'", () => {
+      expect(Validate.user("group:engineering#member")).toBeTruthy();
+    });
+    it("shoud pass for 'UserObject'", () => {
+      expect(Validate.user("group:engineering")).toBeTruthy();
+    });
+    it("shoud pass for 'UserWildcard'", () => {
+      expect(Validate.user("group:*")).toBeTruthy();
```
Comment (rhamzeh) [inline pkg/js/tests/validate-rules.test.ts:23]: Same here, can we add more test cases?

File: pkg/js/tests/validate-rules.test.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      expect(Validate.userObject("group#member")).toBeFalsy();
+    });
+  });
+
+  describe("Rule 'UserWildcard'", () => {
+    it("should pass '<type>:*", () => {
+      expect(Validate.userWildcard("group:*")).toBeTruthy();
+    });
+
+    it("should fail if missing '*' delimiter", () => {
```
Comment (rhamzeh) [inline pkg/js/tests/validate-rules.test.ts:60]: should fail e.g. `user:anne`

And add more test cases like above

File: pkg/js/tests/validate-rules.test.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      expect(Validate.userWildcard("group")).toBeFalsy();
+    });
+  });
+
+  describe("Rule 'Type'", () => {
+    it("should pass '<type>", () => {
+      expect(Validate.type("folder")).toBeTruthy();
```
Comment (rhamzeh) [inline pkg/js/tests/validate-rules.test.ts:67]: Can you add the ones that would fail?

File: pkg/js/validator/validate-rules.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+export const Rules = {
+  type: "[^:#@\\s]{1,254}",
+  relation: "[^:#@\\s]{1,50}",
+  condition: "[^\\s]{2,256}",
+  object: "[^\\s]{2,256}",
+  id: "[^#:\\s]+",
+};
```
Comment (rhamzeh) [inline pkg/js/validator/validate-rules.ts:7]: If we have any validation on the server for the condition param name, let's add it here

File: pkg/js/validator/validate-rules.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+export const Validate = {
+  // An Object is composed of a type and identifier (e.g. 'document:1')
+  object: (object: string): boolean => {
+    return validateFieldValue(`^${Rules.object}:${Rules.id}$`, object);
```
Comment (rhamzeh) [inline pkg/js/validator/validate-rules.ts:12]:
```suggestion
    return validateFieldValue(`^${Rules.type}:${Rules.id}$`, object);
```

This should ALSO pass Rules.object, but Rules.object is for the full thing

File: pkg/js/validator/validate-rules.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  relation: (relation: string): boolean => {
+    return validateFieldValue(`^${Rules.relation}$`, relation);
+  },
+  // User is composed of type, identifier and relation (e.g. 'group:engineering#member')
+  userSet: (user: string): boolean => {
+    return validateFieldValue(`^${Rules.type}:${Rules.id}#${Rules.relation}$`, user);
+  },
+  // User is composed of type and identifier (e.g. 'group:engineering')
+  userObject: (userObject: string): boolean => {
+    return validateFieldValue(`^${Rules.type}:${Rules.id}$`, userObject);
```
Comment (rhamzeh) [inline pkg/js/validator/validate-rules.ts:24]: This should redirect to validate against object

File: pkg/js/validator/validate-store.ts [hunk 8]
```diff
@@ -0,0 +1,976 @@
...
+// Validate Relation
+function validateRelation(tuple: TupleKey, typeDefs: TypeDefinition[], instancePath: string): boolean {
+  const errors = [];
+
+  // Check if relation exists on given type
+  let doesExistOnType = false;
+  if (tuple.user.includes("#")) {
+    const user = tuple.user.split(":")[0];
+    const userRelation = tuple.user.split("#")[1];
+    for (const typeDef of typeDefs) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:203]: In a follow-up PR, let's pre-compute these. The language representation of the model should have all the relations per type precomputed so as to improve performance.

File: pkg/js/validator/validate-store.ts [hunk 9]
```diff
@@ -0,0 +1,976 @@
...
+function validateTypeRestrictions(
+  tuple: TupleKey,
+  typeDefs: TypeDefinition[],
+  conditions: Record<string, Condition> | undefined,
+  instancePath: string,
+): boolean {
+  validateTuple.errors = validateTuple.errors || [];
+
+  const mappedTuple = mapTuple(tuple);
+  const object = tuple.object.split(":")[0];
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:259]: What is this doing? It seems to be getting the type, not the object

File: pkg/js/validator/validate-store.ts [hunk 10]
```diff
@@ -0,0 +1,976 @@
...
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Validation for tuples
+const validateTuple: SchemaValidateFunction = function (
+  this: { jsonModel: AuthorizationModel },
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:324]: Any particular reason you are using `this` as the variable name?

File: pkg/js/validator/validate-store.ts [hunk 11]
```diff
@@ -0,0 +1,976 @@
...
+  tuple: TupleKey,
+  cxt: { instancePath: string },
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:326]:
```suggestion
  ctx: { instancePath: string },
```

File: pkg/js/validator/validate-store.ts [hunk 12]
```diff
@@ -0,0 +1,976 @@
...
+    validateRelation(tuple, jsonModel.type_definitions, cxt.instancePath) &&
+    validateConditionExists(tuple, jsonModel.conditions, cxt.instancePath) &&
+    validateTypeRestrictions(tuple, jsonModel.type_definitions, jsonModel.conditions, cxt.instancePath)
+  );
+};
+
+// Validate the user field for check and list_objects
+function validateUserField(model: AuthorizationModel, types: string[], userField: string, instancePath: string) {
+  const errors = [];
+  const user = userField.split(":")[0];
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:353]: This is also getting the type, not the object

File: pkg/js/validator/validate-store.ts [hunk 13]
```diff
@@ -0,0 +1,976 @@
...
+  const context = listObjects.context;
+  for (const testParam in context) {
+    if (!params.includes(testParam)) {
+      errors.push(unidentifiedTestParam(testParam, instancePath + `/context/${testParam}`));
+    }
+  }
+  return errors;
+}
+
+// Validate List User
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:509]: List Users

File: pkg/js/validator/validate-store.ts [hunk 14]
```diff
@@ -0,0 +1,976 @@
...
+  instancePath: string,
+) {
+  const errors = [];
+
+  const types = model.type_definitions.map((d) => d.type);
+
+  if (listUsers && isStringValue(listUsers.object)) {
+    const listUserObj = listUsers.object;
+
+    const object = listUserObj.split(":")[0];
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:524]:
```suggestion
    const type = listUserObj.split(":")[0];
```

Comment (rhamzeh) [reply pkg/js/validator/validate-store.ts:524]: Considering this is used everywhere, can we make this a helper?

Have a tuple parser that converts from a string to the required fields, so that this and similar calls can become: `const { type } = parseObjectString(listUserObj)`


And `parseObjectString` would return `{ type, id }`
And `parseUserString` would return `{ type, id, relation, wildcard }`

File: pkg/js/validator/validate-store.ts [hunk 15]
```diff
@@ -0,0 +1,976 @@
...
+    if (!errors.length) {
+      if (!tuples.map((tuple) => tuple.object).filter((object) => object === listUserObj).length) {
+        errors.push(undefinedTypeTuple(listUserObj, instancePath + "/object"));
+      }
+    }
+  }
+
+  // Check user fileter
+  if (listUsers.user_filter) {
+    for (const typeNo in listUsers.user_filter) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:540]: Can you call it userFilter instead? I'm not sure what typeNo is an abbreviation of

File: pkg/js/validator/validate-store.ts [hunk 16]
```diff
@@ -0,0 +1,976 @@
...
+
+  // Collect params for validity checking
+  const params: string[] = [];
+  for (const condition in model.conditions) {
+    for (const param in model.conditions[condition].parameters) {
+      params.push(param);
+    }
+  }
+
+  for (const testNo in store.tests) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:609]: call it singleTest instead of testNo

File: pkg/js/validator/validate-store.ts [hunk 17]
```diff
@@ -0,0 +1,976 @@
...
+    if (!test) {
+      continue;
+    }
+
+    if (test.tuples && test.tuples.length) {
+      tuples.push(...test.tuples);
+    }
+
+    // Validate check
+    for (const checkNo in test.check) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:626]:
```suggestion
    for (const singleCheckTest of test.check) {
```

File: pkg/js/validator/validate-store.ts [hunk 18]
```diff
@@ -0,0 +1,976 @@
...
+
+  for (const testNo in store.tests) {
+    // Collect valid tuples
+    const tuples = [];
+    if (store.tuples && store.tuples.length) {
+      tuples.push(...store.tuples);
+    }
+
+    const test = store.tests[testNo];
+    if (!test) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:617]: How can this be false? You're looping over the store tests already

File: pkg/js/validator/validate-store.ts [hunk 19]
```diff
@@ -0,0 +1,976 @@
...
+          test.check[checkNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/check/${checkNo}`,
+        ),
+      );
+    }
+
+    // Validate list objects
+    for (const listNo in test.list_objects) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:642]:
```suggestion
    for (const singleListObjectsTest of test.list_objects) {
```

Comment (rhamzeh) [reply pkg/js/validator/validate-store.ts:642]: That way you don't need to do `if !test.list_objects[listNo]` below, just
```
if !singleListObjectsTest.user || !singleListObjectsTest.type)
```

File: pkg/js/validator/validate-store.ts (hunk 17 above)
Comment (rhamzeh) [reply pkg/js/validator/validate-store.ts:626]: With the appropriate changes below

File: pkg/js/validator/validate-store.ts [hunk 20]
```diff
@@ -0,0 +1,976 @@
...
+          model,
+          test.list_objects[listNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/list_objects/${listNo}`,
+        ),
+      );
+    }
+
+    for (const listNo in test.list_users) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:657]: Same comment as above, use `for of` + singleListUsersTest

File: pkg/js/validator/validate-store.ts [hunk 21]
```diff
@@ -0,0 +1,976 @@
...
+
+  if (errors.length) {
+    validateStore.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+const validateStore: SchemaValidateFunction = function (
+  this: { jsonModel: AuthorizationModel },
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:681]: this and ctx, can we have better names?

File: pkg/js/validator/validate-store.ts [hunk 22]
```diff
@@ -0,0 +1,976 @@
...
+      type: "object",
+      schema: false,
+      errors: true,
+      validate: validateTuple,
+    })
+    .compile(OPENFGA_TUPLES_SCHEMA);
+}
+
+// YAML tuple validation
+const OPENFGA_TUPLES_SCHEMA: Schema = {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-store.ts:727]: You might want to have this in a separate file if you want to re-use this across go, js and Java. But you can move it out in a separate PR

File: pkg/js/tests/validate-rules.test.ts (hunk 1 above)
Comment (d-jeffery) [reply pkg/js/tests/validate-rules.test.ts:10]: Separating test cases into 2 groups, those specific to a validator, and those which are just checks for bad formats that should universally fail.

File: pkg/js/validator/validate-store.ts (hunk 10 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:324]: It's an argument passed in during invocation of `.call()`

https://github.com/openfga/language/pull/286/files#diff-545133f119b2a360e2aa4a2c410f1a5e39fe617ab5377005c29f96f42ce61ae9R32

This was taken from documentation I read when I originally wrote it. I'll try to simplify and remove the `this` assignment.

File: pkg/js/validator/validate-store.ts (hunk 9 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:259]: Renamed variables.

File: pkg/js/validator/validate-store.ts (hunk 8 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:203]: https://github.com/openfga/language/issues/290

File: pkg/js/validator/validate-store.ts (hunk 11 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:326]: I've changed over the names - `cxt` is the parameter name from ajv. 

https://ajv.js.org/keywords.html#schema-compilation-context

File: pkg/js/validator/validate-store.ts (hunk 15 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:540]: Renamed

File: pkg/js/validator/validate-store.ts (hunk 18 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:617]: Removed.

File: pkg/js/validator/validate-store.ts (hunk 19 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:642]: Problem with `of` is that it throws a `is not iterable` exception on **undefined** values, where as `in` doesn't.

Happy to make changes to use `of`, I'll just have to null-check ahead of the loop - is that preferable for you?

File: pkg/js/validator/validate-store.ts (hunk 12 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:353]: Renamed

File: pkg/js/validator/validate-store.ts (hunk 21 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-store.ts:683]: `jsonModel` is needed for both the validation of the overall store, as well as the tuples.

I'm setting `this` here, via `call`, such that the `jsonModel` is persisted for all validation being referenced by `validateStore` & `validateTuples`. 

Doing so means we dont have to rebuild the schema validator every time we want to invoke  the validation.

I've tried today to do this several alternatives today, and I cant find a way to provide the model data.

File: pkg/js/validator/validate-store.ts (hunk 22 above)
Comment (rhamzeh) [reply pkg/js/validator/validate-store.ts:729]: Isn't this just using [JSON Schema](https://json-schema.org/)? That's a standard so should be applicable for Go and Java too.

See https://json-schema.org/implementations

E.g.
Java: https://github.com/networknt/json-schema-validator
Go: https://github.com/kaptinlin/jsonschema

File: pkg/js/validator/validate-rules.ts (hunk 6 above)
Comment (d-jeffery) [reply pkg/js/validator/validate-rules.ts]: Is there a reason to keep `Rules.object` definition then? It appears to only be used to enforce string length, and that it includes no spaces.

Both `Rules.type` and `Rules.id` are more strict and enforce no spacing. Can I just add length check to the strings if that is the case?

Comment (d-jeffery) [reply pkg/js/validator/validate-rules.ts]: Also does the `id` have a max length, or is this just checked by `Rules.object`

**Review by rhamzeh — APPROVED** [review]
FYI I removed the reference to https://github.com/openfga/language/issues/226 from the description
